prefect_saturn/core.py

```diff
@@ -136,91 +148,131 @@ def register_flow_with_saturn(self, flow: Flow) -> bool:
+        return self._saturn_image
 
-        This method sets the `image_tag` to a random string to avoid conflicts.
-        The image name is generated in Saturn.
+    def register_flow_with_saturn(
+        self, flow: Flow, dask_cluster_kwargs=None, dask_adapt_kwargs=None
+    ) -> Flow:
         """
-        storage_details = self.storage_details
-        image_tag = str(uuid.uuid4())
-        # NOTE: SATURN_TOKEN and BASE_URL have to be set to be able
-        #       to load the flow. Those variables will be overridden by
-        #       Kubernetes in all the places where it matters, and values
-        #       set in the image are overridden by those from kubernetes
-        #       https://kubernetes.io/docs/tasks/inject-data-application/define-environment-variable-container
-        flow.storage = Docker(
-            base_image=self._saturn_base_image,
-            image_name=storage_details["image_name"],
-            image_tag=image_tag,
-            registry_url=storage_details["registry_url"],
-            prefect_directory="/tmp",
-            env_vars={"SATURN_TOKEN": "placeholder-token", "BASE_URL": "placeholder-url"},
-            python_dependencies=["kubernetes"],
-            ignore_healthchecks=True,
+        Given a flow, set up all the details needed to run it on
+        a Saturn Dask cluster.
+
+        This method modifies the following components of ``Flow`` objects
+        passed to it.
+
+        * ``.storage``: a ``Webhook`` storage instance is added
+        * ``.environment``: a ``KubernetesJobEnvironment`` with a ``DaskExecutor``
+            is added. This environment will use the same image as the notebook
+            from which this code is run.
+        """
+        self._set_flow_metadata(flow)
+
+        storage = self._get_storage()
+        flow.storage = storage
+
+        environment = self._get_environment(
+            cluster_kwargs=dask_cluster_kwargs, adapt_kwargs=dask_adapt_kwargs
         )
+        flow.environment = environment
+
+        flow.storage.add_flow(flow)
+
+        # tell Saturn to store the flow
+        self._store_flow(flow=flow)
         return flow
 
-    def build_storage(self, flow: Flow) -> Response:
+    def _get_storage(self) -> Webhook:
         """
-        Actually build and push the storage
+        Create a `Webhook` storage object with Saturn-y details.
         """
-        if self._saturn_flow_id is None:
-            raise RuntimeError(Errors.NOT_REGISTERED)
+        storage = Webhook(
+            build_request_kwargs={
+                "url": "${BASE_URL}api/prefect_cloud/flows/" + self.flow_id + "/flow_content",
+                "headers": {
+                    "Content-Type": "application/octet-stream",
+                    "Authorization": "token ${SATURN_TOKEN}",
+                },
+            },
+            build_request_http_method="POST",
+            get_flow_request_kwargs={
+                "url": "${BASE_URL}api/prefect_cloud/flows/"
```

Comment:
Looks like we use the `api/prefect_cloud/flows/:id/flow_content` URL twice in this function. May want to make it a constant and add the `?flow_version_id` bit here.

Reply:
good idea. I just consolidated it into one `url`